### PR TITLE
Speed Memory Tradeoff

### PR DESCRIPTION
**Pull request by igm503**
Note: I'm putting this up here because it's a bit interesting, but not suggesting changes to the repo

I spent a couple of hours trying to see if I could get the benchmarks to run in the main process without having my memory explode, and after that failed, spent some more time trying to figure out why memory usage explodes in the first place, and it seems to be because of some unexpected (to me, at least) behavior on the part of the MLX memory allocator. 

PyTorch's mps and cuda modules have cache clearing functions (e.g. torch.mps.empty_cache()), because unlike what usually happens in Python, the mps and cuda backends will greedily hold onto memory that has been allocated, even if there are no remaining references to a previously allocated tensor. I'm guessing this is because PyTorch often deals with huge amounts of memory, and because quickly giving back unused memory would mean lots of wasted time deallocating and then reallocating it when it's needed again. Eventually they will release the memory, but if you need to immediately clear out the memory that is not being used for currently referenced tensors, you can call torch.BACKEND.empty_cache(), and get the memory back. 

This means that if the benchmark flow is changed from
```
for op in op_list:
    for _ in iterations:
        for framework in frameworks:
            do benchmark
```
to 
```
for framework in frameworks:
    for op in op_list:
        for _ in iterations:
            do benchmark
```
then all the mps benchmarks can be done first, calling torch.mps.empty_cache() after each one, so once the benchmarks are going to be run on mlx, all or most of the torch memory has been deallocated, and mlx won't be competing with torch for memory. The effect, at least on my machine, is that I don't see spikes to >25 gb of memory usage, even though I'm not running each benchmark in a separate process!

However, once the mlx benchmarks begin, there doesn't seem to be anything that can be done to stop the mlx memory allocator from holing up ~15 gigs of memory. I tried for quite a while, however, and found several seemingly general rules:
- removing all references to an mlx array that has been evaluated will *not* on its own cause its memory to be returned to the system
- as far as I can tell, that memory will not, in general, be returned to the system until the program exits
- there is no function equivalent to torch.mps.empty_cache(), so mlx can't be forced to give unused memory back 
- mlx *will not* allocate additional memory if a new array is evaluated that has the shape of a previously evaluated array that no longer has a reference
- mlx *will* allocated additional memory if a new array is evaluated that has a different shape from no longer referenced evaluated arrays, even if it has plenty of memory already in terms of size! (I think this is because memory allocations are done on a per-array basis, so in most cases a different size or data type won't fit)
- mlx will deallocate memory if it grows too big, but it's unclear to me whether this is something handled by mlx or handled by the operating system (it seems to take quite a while before mlx starts to deallocate memory when there's very high memory pressure, which makes me wonder if it's the os stepping in, and not mlx, but again, speculation)
### Example
```
a = mx.random.normal(shape) # say this array takes 1 gig of memory
# 0 gigs allocated, because a hasn't been evaluated

mx.eval(a)
# 1 gig allocated, because a was evaluated

b = mx.random.normal(shape)
# 1 gig allocated

mx.eval(b)
# 2 gigs allocated

a = None
b = None
# 2 gigs allocated--deleting references doesn't free memory

c = mx.random.normal(shape)
mx.eval(c)
# 2 gigs allocated--because c is the same shape (and let's assume dtype) as a and b, 
# mlx doesn't allocate new memory

d = mx.random.normal(shape)
mx.eval(d)
# 2 gigs allocated

e = mx.random.normal(shape)
mx.eval(e)
# 3 gigs allocated--c and d took up the space that was originally allocated for a and b; 
# now more is needed for e

c = None
d = None
e = None
# 3 gigs allocated

f = mx.random.normal(different_shape)
mx.eval(f)
# > 3 gigs allocated--because f is a different shape than any of the previously initialized
# arrays, new memory is allocated, even if f is smaller than the previous arrays!
```
So, it seems like this is why memory usage gets pretty high when running the benchmark on mlx ops. Many different shapes of arrays are created by the benchmark class instances, so even if you delete the references to the input data created by each of those instances, mlx will not free the memory, and will often allocate new memory when an array with a new shape or dtype is created. 

One final note, however, is that although the memory used by mlx gets very high, I haven't seen it crash while running the benchmark in the way described above. When pytorch mps and mlx ops were running interleaved with each other in the main process, sometimes the benchmark would exit because there wasn't enough memory (for me, this was usually pytorch throwing a memory related exception). If they aren't interleaved however, by first doing all the mps tests, clearing the mps cache, and then running all of the mlx tests, mlx is able to run all of the tests fine, albeit at the cost of putting some of your other programs into swap. 

If anyone also finds this interesting and has a macbook with 8 or 16gb of memory, I'd be very curious to see what happens when you run this version of the benchmark on your machine! Wondering in particular what memory usage you get, how much your system swaps, etc.

### CODE REVIEW

**Comment by TristanBilot** [comment]
This is very interesting @igm503, thanks for taking the time to make this analysis. I also tried to use a single process on which running all benchmarks in the first place but it resulted with the same issues. As you mentioned, I noticed that memory consistently grows when the mlx benchmark begins.

Based on your code (i.e. without multiprocessing), I tried to run each of `mlx_gpu` and `mps` first in the benchmark to study the allocated memory of the python process **before** and **after** each benchmark ran.
Here are the results:

`mlx_gpu`
- Before running: 4GB swap
- After running: 11GB Python process memory , 9GB swap

`mps`
- Before running: 5.8GB swap
- After running: 370MB Python process memory , 5.6GB swap

This clearly shows that many artefacts remain in the mlx memory, where `mps` successfully releases its allocated tensors.

When running your code as is, I get no memory error though, and I still have a memory peak of ~11GB, which I suppose to be the blocked memory allocated by mlx.

**Comment by TristanBilot** [comment]
@awni this may be of interest for you. Is this expected behavior?

**Comment by awni** [comment]
This is very interesting, thank you. 

> Is this expected behavior?

Expected yes, optimal no.

**Comment by awni** [comment]
This is the related issue in MLX https://github.com/ml-explore/mlx/issues/742

I don't know yet the path forward yet, but it's a priority to either improve allocation strategy in a general way and/or expose options to make it easier to configure for the given situation.

**Comment by awni** [comment]
> removing all references to an mlx array that has been evaluated will not on its own cause its memory to be returned to the system

Expected, they are held in the cache.

> as far as I can tell, that memory will not, in general, be returned to the system until the program exits

Unless you use a lot of memory, then the cache will be garbage collected

> mlx will not allocate additional memory if a new array is evaluated that has the shape of a previously evaluated array that no longer has a reference

There are conditions for reusing buffers. Basically the buffer has to be big enough, but not too much bigger than the requested size.

> mlx will deallocate memory if it grows too big, but it's unclear to me whether this is something handled by mlx or handled by the operating system

This is handled by MLX

**Comment by TristanBilot** [comment]
> Unless you use a lot of memory, then the cache will be garbage collected

So if I understand well: in theory, we shouldn't have any memory overflow if we run all MLX operations in a single process (unless the allocation is larger than the available memory), as the memory from past allocations will always be freed for future large allocations within the process.
However, we should not run MLX/PyTorch (or other framework) operations together in a single process, as the memory allocated by MLX will not be freed to let the other frameworks allocate their memory within the process.

If this is the case, I guess we can reduce the number of process spawns in the benchmark @igm503. We could even run all framework operations in the main process except MLX operations in a separate process.

**Comment by awni** [comment]
> So if I understand well: in theory, we shouldn't have any memory overflow if we run all MLX operations in a single process (unless the allocation is larger than the available memory), as the memory from past allocations will always be freed for future large allocations within the process.

Correct!

> However, we should not run MLX/PyTorch (or other framework) operations together in a single process, as the memory allocated by MLX will not be freed to let the other frameworks allocate their memory within the process.

Also correct! MLX will hold buffers in the cache and PyTorch has no way to get them from us. It's an issue we are aware of and something to improve.

**Comment by awni** [comment]
I would really like to know if your benchmarks change when you reduce # of processes. I don't think they should..but you never know.

**Comment by igm503** [comment]
> If this is the case, I guess we can reduce the number of process spawns in the benchmark @igm503. We could even run all framework operations in the main process except MLX operations in a separate process.

I'll try it out!

File: mlx_benchmark/run_benchmark.py [hunk 1]
```diff
@@ -23,107 +22,110 @@
...
-                p = mp.Process(target=run, args=(op, args, queue))
-                p.start()
+    print("\nDetailed benchmark:")
+    print_benchmark(all_times, backends)
+    print("\n Average benchmark:")
+    print_benchmark(all_times, backends, reduce_mean=True)
 
-                times = queue.get()
-                p.join()
+def run_mlx_backend(operations, backend, iterations, ops_per_process=10):
```
Comment (igm503) [inline mlx_benchmark/run_benchmark.py:46]: ops_per_process determines the number of ops to run in each process. Decreasing this number decreases the max memory usage.

**Comment by igm503** [comment]
I've put the mlx_gpu, mlx_gpu_compiled, and mlx_cpu benchmarks back into their own processes. To reduce the time spent spawning and cleaning up processes, it's running 10 ops per process, rather than 1 op per process. The max memory usage I'm seeing is about 8 gigs. If you think that's a little high, it can be reduced by decreasing the number of ops per process (running 5 ops per process gives a max of around 4-5 gigs, IIRC).

@awni There doesn't appear to be a difference between running in a separate process and not, but it looks like 2d convolutions are faster in 0.5.0!

**Comment by awni** [comment]
> There doesn't appear to be a difference between running in a separate process and not, but it looks like 2d convolutions are faster in 0.5.0!

Awesome, thanks for checking!!! And yea our convolutions got (sometimes a lot) faster in 0.5!

Also you may be interested in this PR https://github.com/ml-explore/mlx/pull/761. Adds some memory use apis to help understand and control memory / cache behavior in MLX.

**Comment by TristanBilot** [comment]
Amazing work @igm503, thanks! It seems the benchmark is now running faster without these many processes. 🚢

**Pull request merged**